pkg/controller/elasticsearch/elasticsearch_controller.go

```diff
@@ -187,6 +187,8 @@ func (r *ReconcileElasticsearch) Reconcile(request reconcile.Request) (reconcile
 	if err != nil {
 		if apierrors.IsNotFound(err) {
 			// Object not found, return.  Created objects are automatically garbage collected.
+			// Stop tracking that cluster in expectations - without the finalizer overhead.
```

Comment:
Finalizers are generally better since we have a guarantee that things are going to be cleaned up eventually. If we did set a finalizer here instead, the Elasticsearch resource would not be deleted until the operator cleans its own expectations state.
However that also means that if the operator is not running, you cannot delete the ES resource.

It makes a lot of sense when interacting with external systems (ie. not in-memory stuff), since you still want to make sure you clean things up eventually, even though it means waiting for the operator to be running.
In our case I think it does not matter, and it looks like the [apiserver/client-go implementation](https://github.com/kubernetes-sigs/controller-runtime/pull/88#issuecomment-408533747) is robust enough to never miss a deletion event (I guess that cannot be proved though?). So we either eventually get the deletion event, or restart in which case we may have missed the event but we have nothing in memory anyway.

I tend to think we could get rid of most (all?) finalizers for that reason. I'll create another issue.
If you think it's better for consistency in the current codebase I can replace this by a finalizer (until we maybe get rid of them all).

Reply:
I think it's reasonable to leave it as it is now, thanks for the explanation 👍